### DISCUSSION TITLE
Stateful variables (streams, live variables, reactivity, ... whatever we call it)

### DISCUSSION THREAD

**Discussion by dumblob** [uwu-tech/Kind#234] [General]
What could be the most convenient way to model something like https://github.com/MaiaVictor/PureState in Kind?

I'm starting to tinker with Kind a bit, but I'm still too far away from writing anything bigger than hello world :wink:.

**Comment by VictorTaelin** [comment]
Imagine a runtime like Haskell's STG, except that it allowed mutable variables. Of course, if you mutate a variable, then you break purity and referential transparency... but **not** if you "notify" every thunk or expression that depended on the mutable variable, and then re-compute their values. It would be as if a physics constant changed and the whole universe was re-computed to consider it. Something like that would allow, say, react-like UI management without needing to compute diffs. I think it is a fun concept and I could see a runtime based on it. It can also probably be easily modeled in Kind, using monads to have a nice syntax for it. But other than that I don't think there is much else to say about it. I don't see any interesting / compelling use other than UIs.

**Comment by dumblob** [comment]
>It can also probably be easily modeled in Kind, using monads to have a nice syntax for it.

Do you have anything specific in mind?

>But other than that I don't think there is much else to say about it. I don't see any interesting / compelling use other than UIs.

Yep, it's for UI of all kind. But UIs are important - even `printf()` is UI, so I feel this concept be quite useful in most apps :wink:. Such "live" UIs (including distant collaboration!) are yet to become ubiquitous IMHO.

**Comment by VictorTaelin** [comment]
> Do you have anything specific in mind?

Not in mind, just a guess.

> Yep, it's for UI of all kind. But UIs are important - even printf() is UI, so I feel this concept be quite useful in most apps 😉. Such "live" UIs (including distant collaboration!) are yet to become ubiquitous IMHO.

Fair enough, but when it comes to UIs, we're currently betting on React-like interfaces. Check the apps on http://uwu.tech/ currently. So observables isn't something I plan to invest on anytime soon.

**Comment by dumblob** [comment]
>Fair enough, but when it comes to UIs, we're currently betting on React-like interfaces.

Sounds rather heavy-weight and not so quick to develop & maintain. Have you also considered a bit lighter-weight paradigms? E.g. something like:
1. https://github.com/adamhaile/S (like PureState but minimizing the "problem with conditions")
2. https://github.com/WebReflection/domdiff (I usually prefer this one among other options here)
3. https://github.com/ajnsit/concur-js (inspiration in Haskel concur)
4. https://github.com/ms-jpq/noact (tiny React)
5. https://github.com/sweetpalma/gooact (tiny React)
6. https://github.com/zserge/o (tiny React)
7. https://github.com/paldepind/flyd (explicit streams - quite verbose, but it gets the job done)

It can also be seen as a form of [pub sub pattern](https://en.wikipedia.org/wiki/Publish%E2%80%93subscribe_pattern ) - so even https://github.com/moothyknight/JS_UI_Utils would do the job in the end.

>Check the apps on http://uwu.tech/ currently.

Looks good to me (assuming the size of the generated JS files will get smaller at some point).

**Comment by dumblob** [comment]
As a side note, there is also a whole language which has all variables "live" - see https://github.com/mech-lang (it's based on what the [Eve](http://witheve.com/ ) language achieved - the [video](https://youtu.be/TWAMr72VaaU ) shows what such "live" functional-declarativeness offers - and it's not negligible!).

**Comment by ken-okabe** [comment]
@VictorTaelin 
>Of course, if you mutate a variable, then you break purity and referential transparency... but not if you "notify" every thunk or expression that depended on the mutable variable, and then re-compute their values. 

Actually, this is the knowledge that most programming communities don't weigh much. That is rarely mentioned and I share the idea. This is basically FRP and composes [Directed async graph(DAG)](https://en.wikipedia.org/wiki/Directed_acyclic_graph)
and actually, that's exactly SpreadSheetApplications such as Excel does -> **"notify" every thunk or expression that depended on the mutable variable, and then re-compute their values.**


![image](https://user-images.githubusercontent.com/1316994/152314773-4d2cb05f-44de-4867-9134-8eac891e8157.png)
![image](https://user-images.githubusercontent.com/1316994/152314830-ca58007a-2b49-46cf-a247-cf974ae36353.png)



>It can also probably be easily modeled in Kind, using monads to have a nice syntax for it.

@dumblob
>Do you have anything specific in mind?

Yes, I have implemented in TypeScript.
It's reactive-monad. The minimal FRP in the world.

https://codesandbox.io/s/demo0-helloio-2pvr8?from-embed
https://codesandbox.io/s/demo01-helloio2-537xd?from-embed
https://codesandbox.io/s/demo1-equation-brfh1?from-embed
https://codesandbox.io/s/demo2-timer-qq1vs?from-embed
https://codesandbox.io/s/demo3-canvas-mfvx4?from-embed
https://codesandbox.io/s/demo51-react-hello-431so?from-embed
https://codesandbox.io/s/demo52-react-hooks-ixygc?from-embed
https://codesandbox.io/s/demo55-react-frp-monad-2zve2?from-embed
https://codesandbox.io/s/demo56-react-frp-monad2-77ig4?from-embed
https://codesandbox.io/s/demo-60-petit-dom-hello-2cnec?from-embed
https://codesandbox.io/s/demo-65-petit-dom-frp-monad-zh9ez?from-embed
https://codesandbox.io/s/demo-66-petit-dom-frp-multi-components-6qdpp?from-embed

**Reply by dumblob** [reply]
@stken2050 yes, that looks pure (though somewhat hairy - which is understandable as it's just a showcase of the principle). Thanks for the examples!

I wonder whether Kind is already capable of such interface.

**Reply by ken-okabe** [reply]
> @stken2050 yes, that looks pure (though somewhat hairy - which is understandable as it's just a showcase of the principle). Thanks for the examples!
> 
> I wonder whether Kind is already capable of such interface.

In fact, reactive-monad is my base code for every real-world application. I manage every mutable value as the reactive value. 
Very recently, I tried to implement this to Haskell, but technically they prohibit having a genuine mutable value outside `do STrun` thread, so I found it impossible. That's a really idiotic situation like perhaps Victor has mentioned.
I searched here to confirm if Kind allows the native mutable values.

>looks pure (though somewhat hairy 

I don't understand what you exactly meant. It's basically just an algebraic structure of Functor/Monad.

**Reply by dumblob** [reply]
> I don't understand what you exactly meant. It's basically just an algebraic structure of Functor/Monad.

Yes. I meant it very generally e.g. compared to syntax of "setting variables" (not just in Haskell). I do not think you could make it syntactically more shorter/pleasing in Haskell. Therefore this thread about such stuff in Kind2 :wink: (Kind2 does not have yet finished set of syntax sugar, so there is some opportunity...).

**Comment by ken-okabe** [comment]
@dumblob 
The codesandbox code is transpiled js, and here's the full TypeScript codes.

https://gist.github.com/stken2050/eab59867451a51a9d6a0c454e101651a

Actually, this code is a part of my online programming tutorial book (Japanese and translating to English now).

https://kentutorialbook.github.io/functionalprogramming2022/
